### PR TITLE
Add reference benchmarking flow

### PR DESCRIPTION
**Pull request by jackkoenig**
This makes it easier for developers to benchmark certain changes (I built this for some PRs to come).

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Internal or build-related (includes code refactoring/cleanup)

#### Desired Merge Strategy

- Squash

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by sequencer** [comment]
How about using https://mill-build.com/mill/contrib/jmh.html directly?

**Comment by jackkoenig** [comment]
> How about using https://mill-build.com/mill/contrib/jmh.html directly?

Done.

File: build.sc [hunk 1]
```diff
@@ -15,6 +17,7 @@ object v {
   val scalaCrossVersions = Seq(
     "2.13.12"
   )
+  val scalaVersion = scalaCrossVersions.head
```
Comment (sequencer) [inline build.sc:20]:
```suggestion
  val scalaVersion = scalaCrossVersions.head
  val jmhCoreVersion = "1.37"
```

File: build.sc [hunk 2]
```diff
@@ -304,3 +307,10 @@ trait Lit
   def chiselLitDir: T[os.Path] = T(millSourcePath)
   def litConfigIn: T[PathRef] = T.source(millSourcePath / "tests" / "lit.site.cfg.py.in")
 }
+
+object benchmark extends ScalaModule with JmhModule {
+  def scalaVersion = v.scalaVersion
+  def jmhCoreVersion = "1.37"
```
Comment (sequencer) [inline build.sc:313]: How about put it to `object v`, which simplify the bump flow?
```suggestion
  def jmhCoreVersion = v.jmhCore
```

**Review by sequencer — COMMENTED** [review]
small nitpick;p

File: benchmark/src/main/scala/chisel3/benchmark/Benchmark.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  // Typically 10 iterations for both warmup and measurement is better
+  @Benchmark
+  @OutputTimeUnit(TimeUnit.MICROSECONDS)
+  @Warmup(iterations = 3)
+  @Measurement(iterations = 3)
+  @Fork(value = 1)
+  @Threads(value = 1)
+  def BitPatFromUInt(blackHole: Blackhole): BitPat = {
+    val x = BitPat(0xdeadbeefL.U)
+    // Blackhold consuming the value prevents certain JVM optimizations
```
Comment (seldridge) [inline benchmark/src/main/scala/chisel3/benchmark/Benchmark.scala:25]:
```suggestion
    // Blackhole consuming the value prevents certain JVM optimizations
```

**Review by seldridge — APPROVED** [review]
LGTM

**Pull request merged**